### PR TITLE
Save graph runtime optimizations for minimal build

### PR DESCRIPTION
**Pull request by edgchen1**
**Description**
Add support for saving graph runtime optimizations in an ORT format model. The idea is to allow some optimizations to be "replayed" at runtime in a minimal build. The replaying part will be in another PR.

**Motivation and Context**
Work towards enabling graph optimization in minimal builds.

### CODE REVIEW

File: onnxruntime/core/optimizer/selectors_actions/actions.cc [hunk 1]
```diff
@@ -51,7 +53,7 @@ Status RemoveNodes::Run(Graph& graph, const NodesToOptimize& selected_nodes) con
 }
 
 Status MergeIntoTarget::Run(Graph& graph, const NodesToOptimize& selected_nodes) const {
-  ORT_RETURN_IF_ERROR(MoveInputOutput(graph, selected_nodes, selected_nodes.Target(), value_moves_));
+  ORT_RETURN_IF_ERROR(MoveInputOutput(graph, selected_nodes, selected_nodes.Target(), value_moves_, false));
```
Comment (yuslepukhin) [inline onnxruntime/core/optimizer/selectors_actions/actions.cc:56]: >false [](http://example.com/codeflow?start=100&length=5)

Perhaps, make it a constexpr named var.

File: onnxruntime/core/optimizer/selectors_actions/selector_action_transformer.h [hunk 2]
```diff
@@ -112,7 +119,8 @@ class SelectorActionTransformer : public GraphTransformer {
   Status MatchAndProcess(Graph& graph, Node& node, bool& modified, const logging::Logger& logger) const;
 
   std::unordered_map<std::string, const SelectorAndAction*> op_type_to_selector_and_action_;
-  bool save_;  // save the node groups for use in runtime optimization in a minimal build with an ORT format model
+  // If set, save runtime optimization to graph. Otherwise, apply optimization to graph nodes.
+  optional<RuntimeOptimizationSaveContext> runtime_optimization_save_context_;
```
Comment (yuslepukhin) [inline onnxruntime/core/optimizer/selectors_actions/selector_action_transformer.h:123]: >optional [](http://example.com/codeflow?start=2&length=8)

std:: everywhere?

File: onnxruntime/core/optimizer/selectors_actions/actions.cc (hunk 1 above)
Comment (edgchen1) [reply onnxruntime/core/optimizer/selectors_actions/actions.cc]: added parameter name in comment

File: onnxruntime/core/graph/runtime_optimization_record.h [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+  int num_variadic_outputs;
+};
+
+/** Information for a single runtime optimization.
+It does not contain information about the optimizer itself, that should be maintained seperately.
+*/
+struct RuntimeOptimizationRecord {
+  /** The optimization selector-action identifier. */
+  std::string selector_action_id;
+  /** The nodes to consider for optimization. */
```
Comment (skottmckay) [inline onnxruntime/core/graph/runtime_optimization_record.h:39]: should we call this `action_id` so it's slightly more generic?

File: onnxruntime/core/optimizer/qdq_transformer/selectors_actions/qdq_actions.h [hunk 4]
```diff
@@ -20,6 +20,12 @@ struct QDQReplaceWithNew : public ReplaceWithNew {
       : ReplaceWithNew{domain, op_name, std::move(value_moves)} {}
 
   Status Run(Graph&, const NodesToOptimize& selected_nodes) const override;
+
+#if !defined(ORT_MINIMAL_BUILD)
+  Status RunForSave(Graph& graph, const NodesToOptimize& selected_nodes,
+                    const RuntimeOptimizationSaveContext& save_context,
+                    SavedState& saved_state, bool& graph_modified) const override;
+#endif  // !defined(ORT_MINIMAL_BUILD)
```
[the remark this answers is not in the document]
Comment (skottmckay) [reply onnxruntime/core/optimizer/qdq_transformer/selectors_actions/qdq_actions.h:28]: Correct - however to save those higher level optimizations you need to let at least some of them modify the graph. You don't want those modifications to be captured in the saved graph - only in the runtime optimization info.

File: onnxruntime/core/flatbuffers/schema/ort.fbs [hunk 5]
```diff
@@ -180,6 +180,28 @@ table Attribute{
   graphs:[Graph];
 }
 
+// runtime optimizations
+table NodesToOptimizeIndexes{
+  node_indexes:[uint32];
```
[the remark this answers is not in the document]
Comment (skottmckay) [reply onnxruntime/core/flatbuffers/schema/ort.fbs:185]: wfm

**Comment by edgchen1** [comment]
>   // The version number is defined as kOrtModelVersion in <repo root>/onnxruntime/core/session/inference_session.cc

update, it got moved

---
Refers to: onnxruntime/core/flatbuffers/schema/ort.fbs:260 in eb99cdf. [](commit_id = eb99cdf9e47709e3d269296e4cd36dd7e8c25960, deletion_comment = False)

File: onnxruntime/core/optimizer/qdq_transformer/selectors_actions/qdq_actions.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (edgchen1) [reply onnxruntime/core/optimizer/qdq_transformer/selectors_actions/qdq_actions.h:28]: decided to address this later if needed. added a comment in runtime_optimization_record.h

File: onnxruntime/core/graph/runtime_optimization_record.h [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+ * 2. There is an intermediate state between a runtime optimization and a later runtime optimization that depends on a
+ *    node produced by the first. The ORT format model should contain the initial (not intermediate) graph state plus
+ *    the runtime optimizations. There needs to be a way to make the result of a runtime optimization visible to later
+ *    runtime optimizations and also save the initial graph state.
+ */
+
+namespace onnxruntime {
+
+/** Struct to serialize the node indices in an ORT format model.
+Use NodesToOptimize::EmptyNodeIndex for nullptr entries in the vectors for missing optional inputs
```
Comment (gwang-msft) [inline onnxruntime/core/graph/runtime_optimization_record.h:37]: NodesToOptimize::EmptyNodeIndex is std::numeric_limits<NodeIndex>::max() which is the max of size_t, which is platform dependent

File: onnxruntime/core/graph/runtime_optimization_record_container.cc [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+  if (std::find(optimizations.begin(), optimizations.end(), runtime_optimization_record) != optimizations.end()) {
+    return false;
+  }
+
+  optimizations.emplace_back(std::move(runtime_optimization_record));
+  return true;
+}
+#endif
+
+std::vector<RuntimeOptimizationRecord> RuntimeOptimizationRecordContainer::RemoveRecordsForKey(
```
[the remark this answers is not in the document]
Comment (skottmckay) [reply onnxruntime/core/graph/runtime_optimization_record_container.cc:25]: was more the binary size cost of having individual calls remove a record, but I guess that limited to one call to 'erase'

File: onnxruntime/core/graph/runtime_optimization_record.h (hunk 6 above)
Comment (skottmckay) [reply onnxruntime/core/graph/runtime_optimization_record.h:37]: Should we use std::numeric_limits<uint32_t>::max() as we serialize NodeIndex to the flatbuffer using uint32?

File: onnxruntime/core/graph/runtime_optimization_record_container.h [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+#include <unordered_map>
+#include <vector>
+
+#include "core/common/common.h"
+#include "core/graph/runtime_optimization_record.h"
+
+#if !defined(ORT_MINIMAL_BUILD)
+#define ORT_ENABLE_ADDING_RUNTIME_OPTIMIZATION_RECORDS
+#endif  // !defined(ORT_MINIMAL_BUILD)
+
```
Comment (skottmckay) [inline onnxruntime/core/graph/runtime_optimization_record_container.h:16]: Should the rest of this file be excluded if runtime type optimization is disabled (either via cmake or ifdefs)? Not sure if it's able to be thrown away by the linker so there's no binary size impact in a minimal build if that is the case.

File: onnxruntime/core/graph/runtime_optimization_record_container.cc (hunk 7 above)
[the remark this answers is not in the document]
Comment (edgchen1) [reply onnxruntime/core/graph/runtime_optimization_record_container.cc]: removed it for now. will add the functionality when needed

File: onnxruntime/core/graph/runtime_optimization_record.h (hunk 6 above)
Comment (edgchen1) [reply onnxruntime/core/graph/runtime_optimization_record.h]: changed the value to uint32_t max. we'll have to be careful about not storing larger NodeIndex values in these structures

File: onnxruntime/core/optimizer/selectors_actions/actions.h [hunk 9]
```diff
@@ -16,9 +18,26 @@ class Node;
 
 // actions that are applied to a set of nodes identified during selection
 struct Action {
-  virtual Status Run(Graph&, const NodesToOptimize& selected_nodes) const = 0;
+  struct SavedState {
+    std::vector<uint64_t> produced_node_kernel_def_hashes;
```
[the remark this answers is not in the document]
Comment (edgchen1) [reply onnxruntime/core/optimizer/selectors_actions/actions.h]: WontFix yet

**Review by skottmckay — APPROVED** [review]
:shipit:

**Pull request merged**